scipy/optimize/cython_optimize/__init__.py

```diff
@@ -0,0 +1,138 @@
+with any extra parameters as the 2nd argument. ::
+
+    double (*callback_type)(double, void*)
+
+Import the module into Cython as follows::
+
+    from scipy.optimize.cython_optimize cimport zeros
+
+
+Examples
+--------
+Usage of ``scipy.optimize.cython_optimize`` requires Cython to write callbacks
+that are compiled into C. For more information on compiling Cython see the
+`Cython Documentation <http://docs.cython.org/en/latest/index.html>`_.
+
+These are the basic steps:
+
+1. Create a Cython ``.pyx`` file
+2. Import the ``cython_optimize.zeros`` module
+3. Write the callback function, and call the selected zeros function passing
+   the callback, any extra arguments, and the other solver parameters ::
+
+       import math
+       from scipy.optimize.cython_optimize cimport zeros
+
+       ARGS = {'C0': 1.0, 'C1': 0.7}  # a dictionary of extra arguments
+       XLO, XHI = 0.5, 1.0  # lower and upper search boundaries
+       XTOL, RTOL, MITR = 1e-3, 1e-3, 10  # other solver parameters
+
+       # user defined struct for extra parameters
+       ctypedef struct test_params:
+           double C0
+           double C1
+
+
+       # user defined callback
+       cdef double f(double x, void *args):
+           cdef test_params *myargs = <test_params *> args
+           return myargs.C0 - math.exp(-(x - myargs.C1))
+
+
+       # Cython wrapper function
+       cdef double brentq_wrapper_example(dict args, double xa, double xb,
+                                          double xtol, double rtol, int mitr):
+           # Cython automatically casts dictionary to struct
+           cdef test_params myargs = args  
+           return zeros.brentq(
+               f, xa, xb, <test_params *> &myargs, xtol, rtol, mitr, NULL)
+
+
+       # Python function
+       def brentq_example(args=ARGS, xa=XLO, xb=XHI, xtol=XTOL, rtol=RTOL,
+                          mitr=MITR):
+           '''Calls Cython wrapper from Python.'''
+           return brentq_wrapper_example(args, xa, xb, xtol, rtol, mitr)
+
+
+       x = brentq_example()
+       # 0.6999942848231314
+
+4. If you want to call your function from Python, create a Cython wrapper, and
+   a Python function that calls the wrapper.
+5. Create a Cython ``.pxd`` file if you need to export any Cython functions
+
+
+Full Output
+-----------
+The  functions in ``scipy.optimize.cython_optimize`` can also copy the full
+output from the solver to a C ``struct`` that is passed as its last argument.
+If you don't want the full output just pass ``NULL``. The full output
+``struct`` should contain the following:
+
+- ``funcalls``: number of function calls
+- ``iterations``: number of iterations
+- ``error_num``: error number
+
+An error number of -1 means a sign error, -2 means a convergence error, and 0
+means the solver converged. Note that the full output ``struct`` must be cast
+to ``scipy_zeros_parameters``. Continuing from the previous example::
```

Comment:
This is confusing - this structure comes back in the implementation but should not be public (right?).

Reply:
I also don't understand why the cast `<zeros.scipy_zeros_parameters *> &full_output`  below is necessary.

Reply:
The reason is because currently the `full_output` struct is not rigidly defined, it allows the user to add additional fields, as long as it contain `funcalls`, `iterations`, and `error_num`. But perhaps that's the wrong approach? I can think of two alternatives that would remove the casting requirement:

1. Make the `full_output` struct rigidly defined, and don't allow the end_user to define their own struct, make them use `scipy_zeros_parameters` only.

2. Change the output of the root finders exposed in `cython_optimize` from `double` to `scipy_zeros_parameters`, and remove the `full_output` option.

AFAICT both still require `scipy_zeros_parameters` or a similar equivalent `typedef` to be exposed from the `cython_optimize` to the end user, and would either require adding `double root` to `scipy_zeros_parameters` or making a new and very similar typedef with just `funcalls`, `iterations`,  `error_num`, and `root`.

IMO both alternatives create redundant typedefs and extra code that's an additional burden on the user then just casting their full output to the expected typedef.

### example 1:
This works but user still has to define their own struct if they want to output a Python dictionary, since `scipy_zeros_parameters` has fields that restrict Cython from converting it to a Python dictionary. Also user has to copy the desired fields out of the full output to their simpler struct if they want to output it as a Python dictionary.

<details>

    # in _zeros.pxd
    ctypedef struct scipy_zeros_parameters:
        int funcalls
        int iterations
        int error_num
        callback_type function
        void* args
        double root  # add this or define new full_output typedef

    # in user example
    cpdef full_output_struct brentq_full_output_example(tuple args):
        """takes a tuple, outputs a dictionary"""
        cdef full_output_struct my_full_output  # previously defined by user
        cdef scipy_zeros_parameters full_output
        cdef extra_params myargs
        myargs.a = args
        my_full_output.root = brentq(
            f_example, XLO, XHI, &myargs, XTOL, RTOL, MITR, &full_output)
        # copy full output contents to simpler struct with only primitive fields
        # so Cython will convert to dictionary
        my_full_output.funcalls = full_output.funcalls
        my_full_output.iterations = full_output.iterations
        my_full_output.error_num = full_output.error_num 
        return my_full_output

</details>

### example 2: 
AFAICT similar issues as example 1. If the user wants to use `cpdef` and Python objects, then they would have to create a simpler struct to output the primitive fields: `funcalls`, `iterations`,  `error_num`, and `root` or `cython_optimize` would have to add a redundant output struct that has just those simpler fields.

<details>

    # in _zeros.pxd with changes from example 1
    cdef scipy_zeros_parameters brentq(callback_type f, double xa, double xb, void* args, double xtol, double rtol, int iter)

    # in _zeros.pyx
    cdef scipy_zeros_parameters brentq(callback_type f, double xa, double xb, void *args, double xtol, double rtol, int iter):
        cdef double root
        cdef scipy_zeros_parameters myparams
        myparams.args = args
        myparams.function = f
        root = c_zeros.brentq(scipy_zeros_functions_func, xa, xb, xtol, rtol, iter, <c_zeros.default_parameters *> &myparams)
        myparams.root = root
        return myparams

    # in user example
    cpdef full_output_struct brentq_full_output_example(tuple args):
        """takes a tuple, outputs a dictionary"""
        cdef full_output_struct my_full_output  # previously defined by user
        cdef scipy_zeros_parameters full_output
        cdef extra_params myargs
        myargs.a = args
       full_output = brentq(
            f_example, XLO, XHI, &myargs, XTOL, RTOL, MITR)
        # copy full output contents to simpler struct with only primitive fields
        # so Cython will convert to dictionary
        my_full_output.funcalls = full_output.funcalls
        my_full_output.iterations = full_output.iterations
        my_full_output.error_num = full_output.error_num
        my_full_output.root = full_output.root
        return my_full_output

</details>

Or we could rewrite everything and take a new approach.

Reply:
I've implemented example 1 and pushing it up momentarily

Reply:
✔️ done

Reply:
I don't really understand the argument. In your current example, you have the user define `scipy_brent_full_output`. That should do the job, no need for also `scipy_zeros_parameters`.

Also regarding naming: having `scipy` in the name of any API object is odd, should be avoided.

Reply:
The function definitions in [`_zeros.pxd`](../../../mikofski/scipy/blob/cython_optimize/scipy/optimize/cython_optimize/_zeros.pxd) expect a pointer to a struct with typedef `scipy_zeros_parameters` which has a few more field than just those returned with full output. Those extra fields are important for extracting the callback and extra args and passing them to the underlying C functions in `Zeros/`. But since they're not primitives, Cython won't conveniently convert it to a dictionary when returning the full output, So if the user wants a dictionary returned for full output a new simpler struct is required.

If the user passes their simplified struct instead of the required struct the  the Cython compiler complains that the argument is not the required type as specified in the definition.

I tried to use void instead of a `scipy_zeros_parameters` but then the function has no idea what is in the struct at compile time, or what size it is. This makes getting it's contents at runtime very tricky. I asked some coworkers, and this was the best we could come up with. But I'm open to suggestions if anyone thinks there's a better or easier way. Thanks!

I'm happy to remove the "scipy_" prefixes. Good idea!